console/input.rst

```diff
@@ -308,4 +308,90 @@ The above code can be simplified as follows because ``false !== null``::
     $yell = ($optionValue !== false);
     $yellLouder = ($optionValue === 'louder');
 
+Adding Argument/Option Value Completion
+---------------------------------------
+
+.. versionadded:: 5.4
+
+    Console completion was introduced in Symfony 5.4.
+
+If :ref:`Console completion is installed <console-completion-setup>`,
+command and option names will be auto completed by the shell. However, you
+can also implement value completion for the input in your commands. For
+instance, you may want to complete all usernames from the database in the
+``name`` argument of your greet command.
+
+To achieve this, override the ``complete()`` method in the command::
+
+    // ...
+    use Symfony\Component\Console\Completion\CompletionInput;
+    use Symfony\Component\Console\Completion\CompletionSuggestions;
+
+    class GreetCommand extends Command
+    {
+        // ...
+
+        public function complete(CompletionInput $input, CompletionSuggestions $suggestions): void
+        {
+            if ($input->mustSuggestArgumentValuesFor('names')) {
+                // the user asks for completion input for the "names" option
+
+                // the value the user already typed, e.g. when typing "app:greet Fa" before
+                // pressing Tab, this will contain "Fa"
+                $currentValue = $input->getCompletionValue();
+
+                // get the list of username names from somewhere (e.g. the database)
+                // you may use $currentValue to filter down the names
+                $availableUsernames = ...;
+
+                // then add the retrieved names as suggested values
+                $suggestions->suggestValues($availableUsernames);
```

Comment:
I miss some comments about the limits of suggestions. Can I return an unlimited number here? Is there a hard limit or a soft recommendation? Thanks!

Reply:
It's unlimited (or at least massive, the shell script automatically filters the suggestions and shows a message like "Display all 3197 possibilities? (y or no)" when there are many suggestions).

I've updated the tip below this example a little bit to better highlight it.